src/clj/coffi/mem.clj

```diff
@@ -1350,3 +1766,523 @@
+     [(list `dotimes ['m n]
+       (list `aset a 'm (generate-deserialize array-type `(+ ~offset (* ~(size-of array-type) ~'m)) segment-source-form)))]
+     [a])))
+
+(defn- generate-deserialize-array-as-array [array-type n offset segment-source-form]
+  (if (coffitype->array-read-fn array-type)
+    (generate-deserialize-array-as-array-bulk array-type n offset segment-source-form) ;bulk-copy available
+    (let [inline-cutoff 32] ;this magic value has been benchmarked, but it may need adjusting for specific architectures
+      (if (< n inline-cutoff)
+        (generate-deserialize-array-as-array-inline array-type n offset segment-source-form)
+        (generate-deserialize-array-as-array-loop array-type n offset segment-source-form)))))
+
+(defn- generate-deserialize-array-as-vector-bulk [array-type n offset segment-source-form]
+  (list `vec (list (coffitype->array-read-fn array-type) segment-source-form n offset)))
+
+(defn- generate-deserialize-array-as-vector-loop [array-type n offset segment-source-form]
+  (let [loop-deserialize (generate-deserialize array-type `(+ ~offset (* ~(size-of array-type) ~'i)) segment-source-form)]
+    `(loop [~'i 0 ~'v (transient [])]
+       (if (< ~'i ~n)
+         (recur (unchecked-inc ~'i) (conj! ~'v ~loop-deserialize))
+         (persistent! ~'v)))))
+
+(defn- generate-deserialize-array-as-vector-inline [array-type n offset segment-source-form]
+  (vec (map #(generate-deserialize array-type (+ offset (* (size-of array-type) %)) segment-source-form) (range n))))
+
+(defn- generate-deserialize-array-as-vector [array-type n offset segment-source-form]
+  (let [inline-cutoff 64] ;this magic value has been benchmarked, but it may need adjusting for specific architectures
+    (if (< n inline-cutoff)
+      (generate-deserialize-array-as-vector-inline array-type n offset segment-source-form)
+      (generate-deserialize-array-as-vector-loop array-type n offset segment-source-form))))
+
+(defmethod generate-deserialize :coffi.mem/array    [[_ array-type n & {:keys [raw?]}] offset segment-source-form]
+  (if raw?
+    (generate-deserialize-array-as-array array-type n offset segment-source-form)
+    (generate-deserialize-array-as-vector array-type n offset segment-source-form)))
+
+(defn- typelist [typename fields]
+  (->>
+   (partition 2 2 (interleave (reductions + 0 (map (comp size-of second) fields)) fields))
+   (filter (fn [[_ [_ field-type]]] (not (and (vector? field-type) (= "padding" (name (first field-type)))))))))
+
+(defn register-new-struct-deserialization [typename [_struct fields]]
+  (let [typelist (typelist typename fields)]
+    (defmethod generate-deserialize typename [_type global-offset segment-source-form]
+      (->> typelist
+           (map-indexed
+            (fn [index [offset [_ field-type]]]
+              (generate-deserialize field-type (+ global-offset offset) segment-source-form)))
+           (cons (symbol (str (name typename) ".")))))))
```

Comment:
I'd prefer if this returned the code for the defmethod to be included in the output of the `defstruct` macro, unless there's some specific reason why it should happen at macroexpansion time.

Same with the `register-new-struct-serialization` further down.

Reply:
the rationale here was:

since we want to emit the inlined (de)serialization for the specific types, we need to register the code for that before we emit it. if there is a defstruct that contains another defstruct defined type as a member, `generate-deserialize` must have an implementation for that referenced type at macroexpansion time or else we cannot emit the code for the referenced type.

that is, unless i'm missing something here.

Reply:
as discussed, for clarity, it will both happen at macroexpansion time, but also be emitted as per [1f5efb0](https://github.com/IGJoshua/coffi/pull/14/commits/1f5efb0982335bc93e4d27cbb5d39b8600aaf47e)